src/solver/README.md

```diff
@@ -0,0 +1,238 @@
+ ┗━┛
+```
+###### Diagrams generated with use of http://marklodato.github.io/js-boxdrawing/
+
+where the horizontal block is a candidate, vertical block is a set of sources possible for each resource, and the arrow represents the index of a resource the iterator is currently evaluating.
+
+With those tools introduced, we can now guide the reader through how the algorithm works.
+But before we do that, it is important to justify writing a custom algorithm in place of existing generic solutions, and explain the two testing strategies which heavily impact the algorithm.
+
+# Existing libraries
+The reader may notice that the concept presented here is known as Cartesian Multi-Product There are ready to use solutions for generating such iterators.
+
+#### Rust
+In Rust, crate [`itertools`](https://crates.io/crates/itertools) provides, [`multi_cartesian_product`](https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.multi_cartesian_product) which can be used like this:
+```rust
+use itertools::Itertools;
+
+let multi_prod = (0..3).map(|i| 0..2)
+    .multi_cartesian_product();
+
+for set in multi_prod {
+    println!("{:?}", set);
+}
+```
+([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6ef231f6b011b234babb0aa3e68b78ab))
+
+#### Python
+In Python, `itertools` package provides a function [`itertools::product`](https://docs.python.org/2/library/itertools.html#itertools.product) which can be used to generate such iterator:
+```python
+import itertools
+
+for set in itertools.product(range(2), repeat=3):
+    print(set)
+```
+
+#### Reasons for a custom algorithm
+
+Unfortunately, the computational complexity of generating all possible sets is growing exponentially, both in the cost of CPU and memory use.
+On a high-end laptop, computing the sets for all possible variations of the example above generates *8* sets and takes only *700 nanoseconds*, but computing the same for four sources and 16 resources (a scenario theoretically possible in Firefox with one language pack and Preferences UI for example) generates over *4 billion* sets and takes over *2 minutes*. 
+
+It becomes crucial to micro-optimize the algorithm and to use [Dynamic Programming](https://en.wikipedia.org/wiki/Dynamic_programming) and [Memoization](https://en.wikipedia.org/wiki/Memoization) techniques to minimize the cost of constructing, storing and retrieving sets.
+
+Second important observation is that in most scenarios the resource exists in only some of the sources, and ability to bail out from a branch of candidates that cannot lead to a solution is crucial for performance of the algorithm.
+
+## Optimizations
+
+The algorithm used here is heavily optimized. For the conservative scenario listed above, where 4 sources and 15 resources are all present in every source, the total time on the reference hardware is cut from *2 minutes* to *24 seconds*, while generating the same *4 billion* sets for a **5x** performance improvement. 
```

Comment:
I think "highly efficient" might be better than "heavily optimized" here, since this is an algorithmic improvement rather than hand-writing assembly or something like that.